### PR TITLE
feat: support Python functions runtime for `init` & `deploy`

### PR DESCRIPTION
**Pull request by Salakar**
### Description

A first pass at adding Python functions runtime support for `init` and `deploy`.

cc @inlined @taeold

### Sample Commands

 - `firebase experiments:enable pythonfunctions`
 - `firebase init functions` (choose python)
 - `firebase deploy --only functions`

### CODE REVIEW

File: src/deploy/functions/runtimes/python/index.ts [hunk 1]
```diff
@@ -133,24 +133,30 @@ class Delegate implements runtimes.RuntimeDelegate {
...
         port: 8081,
       });
-      const killProcess = await this.serveAdmin(port, envs);
+      if (Object.keys(configValues || {}).length) {
+        envs.CLOUD_RUNTIME_CONFIG = JSON.stringify(configValues);
+      }
+      const processEnvs: backend.EnvironmentVariables = { ...envs };
+      if (Object.keys(configValues || {}).length) {
+        processEnvs.CLOUD_RUNTIME_CONFIG = JSON.stringify(configValues);
+      }
```
Comment (inlined) [inline src/deploy/functions/runtimes/python/index.ts:152]: Cloud Runtime Config is a feature only present for v1 runtimes (and unfortunately we're currently forced to support it for JS). We can cut it for v2 functions.

**Review by inlined — APPROVED** [review]
Don't yet see the full picture, but I'll have faith ^_^

File: src/deploy/functions/runtimes/index.ts [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
+import * as python from "./python";
 import * as node from "./node";
 import * as validate from "../validate";
 import { FirebaseError } from "../../../error";
 
 /** Supported runtimes for new Cloud Functions. */
 const RUNTIMES: string[] = ["nodejs10", "nodejs12", "nodejs14", "nodejs16"];
+
+// We prompt for runtime selection on functions init, so these are separated.
+export const PYTHON_RUNTIMES: string[] = ["python37", "python38", "python39"];
```
Comment (Salakar) [inline src/deploy/functions/runtimes/index.ts:13]: TODO remove 37, 38, 39

File: src/deploy/functions/runtimes/python/index.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    public readonly runtime: runtimes.Runtime
+  ) {}
+
+  private modulesDir_ = "";
+
+  async modulesDir(): Promise<string> {
+    if (!this.modulesDir_) {
+      const out = await runWithVirtualEnv(
+        [
+          "python3.9",
```
Comment (Ehesp) [inline src/deploy/functions/runtimes/python/index.ts:83]: Above `LATEST_VERSION` is defined as `python39`, but here it's `python3.9` - just confirming that's intended.

Comment (Salakar) [reply src/deploy/functions/runtimes/python/index.ts:83]: Correct, "python39" is the GCP runtime name, "python3.9" is the actual python binary name

**Review by Ehesp — APPROVED** [review]

File: src/init/features/functions/python.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+const MAIN_TEMPLATE = fs.readFileSync(path.join(TEMPLATE_ROOT, "main.py"), "utf8");
+const REQUIREMENTS_TEMPLATE = fs.readFileSync(path.join(TEMPLATE_ROOT, "requirements.txt"), "utf8");
+const GITIGNORE_TEMPLATE = fs.readFileSync(path.join(TEMPLATE_ROOT, "_gitignore"), "utf8");
+
+/**
+ * Create a Python Firebase Functions project.
+ */
+export async function setup(_setup: unknown, config: Config): Promise<void> {
+  await config.askWriteProjectFile(
+    `${Config.DEFAULT_FUNCTIONS_SOURCE}/requirements.txt`,
```
Comment (jamesdaniels) [inline src/init/features/functions/python.ts:17]: should use `setup.functions.source` rather than default here & in other spots

**Review by jamesdaniels — CHANGES_REQUESTED** [review]

File: src/deploy/functions/runtimes/python/index.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import * as cp from "child_process";
+
+import * as runtimes from "..";
+import * as backend from "../../backend";
+import { logger } from "../../../../logger";
+import * as discovery from "../discovery";
+import { FirebaseError } from "../../../../error";
+import { Build } from "../../build";
+
+export const LATEST_VERSION: runtimes.Runtime = "python310";
```
Comment (jamesdaniels) [inline src/deploy/functions/runtimes/python/index.ts:15]: FYI latest is now python311

Comment (taeold) [reply src/deploy/functions/runtimes/python/index.ts:15]: Is it supported in GCF? Per https://cloud.google.com/functions/docs/concepts/python-runtime, I dont see python311.

Comment (jamesdaniels) [reply src/deploy/functions/runtimes/python/index.ts:15]: Hmm, maybe I jumped the gun I was looking a cs

File: src/deploy/functions/runtimes/python/index.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    // No-op.
+  }
+
+  async serveAdmin(port: number, envs: backend.EnvironmentVariables): Promise<() => Promise<void>> {
+    const modulesDir = await this.modulesDir();
+    const envWithAdminPort = {
+      ...envs,
+      ADMIN_PORT: port.toString(),
+    };
+    const args = ["python3.10", path.join(modulesDir, "private", "serving.py")];
```
Comment (jamesdaniels) [inline src/deploy/functions/runtimes/python/index.ts:115]: probably best not to hardcode here, maybe make an const

File: src/init/features/functions/python.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  );
+  await config.askWriteProjectFile(`${Config.DEFAULT_FUNCTIONS_SOURCE}/main.py`, MAIN_TEMPLATE);
+
+  // Write the latest supported runtime version to the config.
+  config.set("functions.runtime", LATEST_VERSION);
+  // Add python specific ignores to config.
+  config.set("functions.ignore", ["venv", "__pycache__"]);
+  // Setup VENV.
+  await runWithVirtualEnv(
+    ["python3.10", "-m", "venv", "venv"],
```
Comment (jamesdaniels) [inline src/init/features/functions/python.ts:32]: same here, best not to hardcode python3.10, use a shared const

**Pull request closed without merging**